### ISSUE TITLE
Razor not re rendering

### ISSUE THREAD

**Issue by jacobc16** [Facepunch/sbox-issues#6852]
### Describe the bug

UI created with razor does not re render properly if using interfaces and is syntax but only if the code inside the if block is a component. It does not update the property that is passed into the component. The main panel will detect the changes, but not update the component that you pass them into.

### To Reproduce

Minimum example: https://github.com/jacobc16/sbox-razor-bug

Type into text entry and see the UI not update, but if you hover over the text it will update

### Expected behavior

The UI should re render the component when the hashcode changes

### Media/Files

https://github.com/user-attachments/assets/f99599ce-73fa-4ce1-8916-c25406b27e17
![Image](https://github.com/user-attachments/assets/7eac0567-b8ae-445a-90ad-b8476aa8e30c)

**Comment by bub-bl** [comment]
The `GetHashCode` function uses a reference for interfaces/classes and a value for value types. I think the rendering is not updated because the instance of the interface itself has not changed.
In your RazorBug.razor, try this line:
csharp
protected override int BuildHash() => System.HashCode.Combine( Name, ExampleDiv, ExampleDiv.Name ) ;
```
Maybe I'm wrong, but I'm pretty sure that's where your problem comes from

Translated with DeepL.com (free version)

**Comment by jacobc16** [comment]
> The `GetHashCode` function uses a reference for interfaces/classes and a value for value types. I think the rendering is not updated because the instance of the interface itself has not changed. In your RazorBug.razor, try this line:
> 
> protected override int BuildHash() => System.HashCode.Combine( Name, ExampleDiv, ExampleDiv.Name );
> 
> Maybe I'm wrong, but I'm pretty sure that's where your problem comes from

I have tried that before and it make no difference. The main component knows it changed, but does not update the component you pass it into. You can use the values in the interface just fine except in the component you tried to pass them into.

**Comment by jacobc16** [comment]
Can be fixed with the new changes from this issue #4103 Unsure if this should remain an open bug, but I will leave it up to whoever reviews this,

**Comment by solwllms** [comment]
Not sure I see a bug here? I think you'll either want to replace Text.BuildHash to use a hash of the property you're using, like: 
`protected override int BuildHash() => System.HashCode.Combine( ExampleDiv.Name );`

or override object.GetHashCode on the ExampleDiv type to generate a useful hash for that object (right now it'll be a static hash per each instance iirc, hence it not updating the panel here)
```
public class Example : IExample
{
	public string Name { get; set; }

	public Example( string name )
	{
		Name = name;
	}

	public override int GetHashCode()
	{
		return System.HashCode.Combine(Name);
	}
}
```
